abbyy_to_epub3/image_processing.py

```diff
@@ -0,0 +1,127 @@
+# Copyright 2017 Deborah Kaplan
+#
+# This file is part of Abbyy-to-epub3.
+# Source code is available at <https://github.com/deborahgu/abbyy-to-epub3>.
+#
+# Abbyy-to-epub3 is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from PIL import Image
+
+import logging
+import subprocess
+
+
+class ImageProcessor(object):
+    """
+    The image object.
+
+    Can use various image processing libraries via factories.
+    """
+    def __init__(self, debug=False):
+        self.logger = logging.getLogger(__name__)
+        if debug:
+            self.logger.addHandler(logging.StreamHandler())
+            self.logger.setLevel(logging.DEBUG)
+
+
+def factory(type):
+
+    class KakaduProcessor(ImageProcessor):
+        def crop_image(self, origfile, outfile, dim=False, pagedim=False):
+            """
+            Given an image object, save a crop of the entire image.
+            """
+            
+            if dim:
+                # if dimensions are passed, save a crop of the image
+
+                # convert the dimensions we have
+                #   (left, top, right, bottom) in pixels
+                # to the format wanted by kakadu
+                #   "{<top>,<left>},{<height>,<width>}"
+                # as percentages between 0.0 and 1.0
+                # pagedim is passed as (width, height)
+
+                if not pagedim:
+                    self.logger.warning(
+                        "Can't crop in Kakadu without page dimensions"
+                    )
+                    return
+
+                (left, top, right, bottom) = dim
+                (pagewidth, pageheight) = pagedim
+                region_string = "{%s,%s},{%s,%s}" % (
+                    top / pageheight,
+                    left / pagewidth,
+                    (bottom - top) / pageheight,
+                    (right - left) / pagewidth
+                )
+
+                cmd = [
+                    'kdu_expand',
+                    '-region', region_string,
+                    '-i', origfile,
+                    '-o', outfile
+                ]
+
+                try:
+                    subprocess.run(cmd, stdout=subprocess.DEVNULL, check=True)
+                except subprocess.CalledProcessError as e:
+                    self.logger.warning(
+                        "Can't save cropped image: {}".format(e)
+                    )
```

Comment:
this will just pass and return `None` implicitly which is fine, but we may want to explicitly `return`, as done above (as convenience to reader)

Reply:
I did think about that and decided not to, but it's my instinct as well, so I'll do happily at that.

Reply:
For what it's worth I'm still not sure what really should happen on this failure. If an image fails to process, should the book continue to produce, just leaving the image out, producing warning messages? Should the entire book fail? The reason I haven't done any sophisticated exception handling up the process call chain is because I'm not sure what the best behavior is.